### DISCUSSION TITLE
OSPF network_type optimization for P2P links

### DISCUSSION THREAD

**Discussion by jbemmel** [ipspace/netlab#1502] [Q&A]
I noticed that https://github.com/ipspace/netlab/blob/34c05695618c83de4170cf3b37560af5c637c2f0/netsim/modules/_routing.py#L87 sets the OSPF ```network_type``` to "point-to-point" in some cases. In my case, it was VLANs between 2 switches with no hosts connected to them (yet)

It occurred to me that even if there were hosts, there would still be only 2 OSPF speakers on the segment - so could this optimization be broadened? Instead of checking ```len(neighbors)==1``` filter for neighbors that have the corresponding routing protocol enabled?

**Comment by ipspace** [comment]
I don't think that's a good idea for scenarios where someone could turn on OSPF (or IS-IS) on another device (IS-IS labs come to mind ;). Anyway, I don't think I've ever seen OSPF/IS-IS P2P type used on edge segments (even though that would technically make sense).

**Reply by jbemmel** [reply]
That's the thing: With Netlab, you get SVI interfaces with p2p OSPF network_type for VLANs with no hosts. Add hosts, and it becomes a broadcast network

Perhaps instead of checking ```len(neighbors)==1```, we should check for ```link.type=='p2p'```?

**Comment by ipspace** [comment]
Looks like we have another bikeshedding exercise. What's wrong with the **ospf.network_type: broadcast** VLAN attribute?

**Reply by jbemmel** [reply]
Well, for one it's not global so if you have 100 VLANs you need to configure 100x "configure this VLAN as broadcast even though there is only 1 neighbor"

**Reply by ipspace** [reply]
So you'd like to have VLAN groups? Any ideas on how that should be modeled?

**Reply by jbemmel** [reply]
I was thinking it might be easier to add ```network_type``` here: https://github.com/ipspace/netlab/blob/21e5445006a03daa8b58278f451de27b38674077/netsim/modules/ospf.yml#L8

and allow the user to say:
```ospf.network_type: broadcast```
globally (and override where needed)

But we could do VLAN groups just like we have node groups:
```
vlans:
 a:
 b:
 c:
 d:

vlangroups:
  g1: [ a,b,c ]
  g2:
    members: [ g1, d ]
    ospf.network_type: broadcast

**Reply by ipspace** [reply]
I have no problem with making OSPF network_type global and node attribute, but don't complain about the results you'll get ;)

**Reply by ipspace** [reply]
> But we could do VLAN groups just like we have node groups:

How about

````
groups:
  edge:
    type: vlan
    members: [ v1, v2, v3 ]
    ospf.network_type: broadcast
````

That would be pretty cool, and we could do it for other dict-like objects as well (VRFs come to mind)

**Reply by jbemmel** [reply]
```type``` could be inferred from the first member, and then check that the other objects are of the same type

**Reply by ipspace** [reply]
> `type` could be inferred from the first member, and then check that the other objects are of the same type

... and the next microsecond someone would use the same name for a node and a VLAN 😜

**Reply by jbemmel** [reply]
...and we could check for that too (and call it "AI")
